.github/workflows/rubyonrails.yml

```diff
@@ -28,6 +28,12 @@ jobs:
     env:
       RAILS_ENV: test
       DATABASE_URL: "postgres://rails:password@localhost:5432/hostedgpt_test"
+      EMAIL_FEATURE: true
+      PASSWORD_RESET_EMAIL_FEATURE: true
+      EMAIL_FROM: support@the.bot
+      EMAIL_HOST: localhost:3000
+      EMAIL_PROVIDER: postmark
+      POSTMARK_SERVER_API_TOKEN: test-token
```

Comment:
sadly the stub_features and stub_settings don't seem to be able to reload the code that is run in the initializers so I added this back temporarily. I'd like to create a custom Config lib or something that we can use to load settings and features into during init and then mock values of during tests. I hate to make it so complex though. Maybe there's an easier way that I haven't found yet. Will keep digging, but I feel like something like that should come in a separate PR since this has already been open for a while. Thoughts?

Reply:
That makes sense since initializers are so early in the app boot. At least the `Feature` is available in the initializer. I initially thought that was what you were suggesting and ran a quick test myself locally and Feature is definitely working with initializers, so that's good.

I think you're on the right track of basically fixing this within the test environment, but I don't think the CI script is the place to do it because we want someone to be able to pull down the repo locally, run tests, and have them work without needing to set special ENV. I think the right strategy is to somehow we need to change the options.yml defaults just for the test environment.

How about this PR I just tee'd up: https://github.com/AllYourBot/hostedgpt/pull/479

If I'm fully understanding the problem, I think this PR will allow you to simply add `test_override: "some value"` within config.yml to make the test environment work.

Reply:
#479 will definitely fix the issue with the incorrect values. I think the problem that I'm running into is not being able to change the values that are dynamically loaded in the application.rb config file and the initializers after the app has started. This means we can't modify one of the features or settings during a test and check that the dependent values (`application.config.whatever`) is set and more importantly we're unable to test the logic that depends on those dependent values. For instance, some of the password reset logic is dependent on `config.password_reset_token_purpose` existing which is set dynamically based on the password reset feature being enabled. I can stub the feature so that it's enabled, but that doesn't rerun the initializer code that set the `config.password_reset_token_purpose` (and whatever else) in the first place. I might be able to just modify the config value at runtime too, but I don't think that's the real solution either because I would have to correctly duplicate the logic from the initializer code to set up the app config values the same way they would be set automatically when the feature env vars are present. Having defaults for the test env will get us set up to run the test with a specific combination of settings/features, but won't allow us to change those env vars during a test and then expect the init code to run and change our app config correctly.

Eh that's a long explanation. Ultimately I think the solution is to not do any dynamic logic in the initializers and move that to a Config class or something if we need to have dynamic logic for a config value. That way we can stub the Config methods. I'd say we can keep it simple and just avoid any dynamic logic during init, but I'm not sure we can hold that off forever given that we're aiming at letting people set this up with different service providers for email, etc.

Reply:
Yea, I follow you. Basically, now that we can set different defaults for the test environment options.yml than for the dev/production defaults we should be able to get a test suite that can pass. But you’re right that as soon as we want to have automated tests for postmark **and** some other service, now we’re back to needing to change the Feature flags within a test.

You’re right that it probably means some way that we move the conditionals outside of initializers, but the problem is that we’re setting rails.application.config values. I’m not too familiar with the inner workings of those.

But it’s also possible that when we do implement another mail provider, it may be trivially different such that there isn’t much additional to test. After all, we’ll want every mail provider to work 99% the same. 

I’m okay crossing that bridge when we come to it. We can just defer solving this problem in a deeper way.